nixos/modules/virtualisation/neutron.nix

```diff
@@ -0,0 +1,369 @@
+      example = "neutron";
+    };
+    servicePassword = mkOption {
+      default = "neutron";
+      description = "The password of the neutron service user";
+      example = "neutron";
+    };
+
+    dbHost = mkOption {
+      default = "localhost";
+      description = "The location of the database server";
+      example = "localhost";
+    };
+
+    dbPassword = mkOption {
+      default = "neutron";
+      description = "The mysql password";
+      example = "neutron";
+    };
+    
+    endpointPublic = mkOption {
+      type = types.str;
+      default = "localhost";
+      description = ''
+      '';
+    };
+
+    keystoneAdminUsername = mkOption {
+      type = types.str;
+      default = "admin";
+      description = ''
+      '';
+    };
+
+    keystoneAdminPassword = mkOption {
+      type = types.str;
+      default = "admin";
+      description = ''
+      '';
+    };
+
+    keystoneAdminTenant = mkOption {
+      type = types.str;
+      default = "admin";
+      description = ''
+      '';
+    };
+  };
+
+
+  config = mkIf cfg.enableSingleNode {
+    # Note: when changing the default, make it conditional on
+    # ‘system.stateVersion’ to maintain compatibility with existing
+    # systems!
+    virtualisation.neutron.package = mkDefault pkgs.neutron;
+
+    # See https://wiki.openstack.org/wiki/Rootwrap
+    security.sudo.enable = true;
+    security.sudo.extraConfig = ''
+      neutron ALL = (root) NOPASSWD: ${package_set}/bin/neutron-rootwrap ${rootwrapConf} *
+    '';
+
+    users.extraUsers = [{
+      name = "neutron";
+      group = "neutron";
+    }];
+    users.extraGroups = [{
+      name = "neutron";
+    }];
+
+    systemd.services.neutron-server = {
+      description = "OpenStack Neutron Daemon";
+      after = [ "rabbitmq.service" "mysql.service" "network.target"];
+      path = [ package_set pkgs.mysql pkgs.curl pkgs.pythonPackages.keystoneclient pkgs.gawk ];
+      wantedBy = [ "multi-user.target" ];
+      preStart = ''
+        mkdir -p /var/lock/neutron /var/lib/neutron
+        chown neutron:neutron /var/lock/neutron/ /var/lib/neutron
+
+        # TODO: move out of here
+        mysql -u root -N -e "create database neutron;" || true
+        mysql -u root -N -e "GRANT ALL PRIVILEGES ON neutron.* TO 'neutron'@'${cfg.dbHost}' IDENTIFIED BY '${cfg.dbPassword}';"
+        mysql -u root -N -e "GRANT ALL PRIVILEGES ON neutron.* TO 'neutron'@'%' IDENTIFIED BY '${cfg.dbPassword}';"
+
+        # Initialise the database
+        ${package_set}/bin/neutron-db-manage --config-file ${neutronConf} --config-file ${ml2PluginConf} upgrade head
+      '';
+      postStart = ''
+	export OS_AUTH_URL=http://localhost:5000/v2.0
+	export OS_USERNAME=${cfg.keystoneAdminUsername}
+	export OS_PASSWORD=${cfg.keystoneAdminPassword}
+	export OS_TENANT_NAME=${cfg.keystoneAdminTenant}
+
+        # Wait until the keystone is available for use
+        count=0
+        while ! keystone user-get ${cfg.keystoneAdminUsername} > /dev/null
```

Comment:
Here you could do a nice trick. Becasue keystone can hang for whatever reason, you could prefix it with timeout i.e: "timeout 30 keystone...."

Next thing (just to consider) is to use $SECONDS variable instead of count, so you know for how long you will be trying. ($SECONDS gives you seconds since bash started, so you could check condition like : 
```
max=$(($SECONDS+30))
if [ $SECONDS < $max ];then
  break
fi
```
)